### ISSUE TITLE
Upgrade plan for v4?

### ISSUE THREAD

**Issue by liayn** [TYPO3/CmsComposerInstallers#131]
I just stumbled over the new stuff in v4 via Bennis blog post.

As v11 allows for v4 already, I fear this will raise quite some unhappy voices. I suspect that most updates will involve the `-W` flag (which we recommend, right?). Hence pipelines, local setups, etc will most probably pull v4 rather "unnoticed", until the instance is corrupted. I assume this will be breaking for a lot of instances. (There is always something hidden the integrator is not aware of in the first place.)

Do you see any way to

a) make this possible breakage very well communicated upfront?
b) make this totally opt-in in v11? (publish as separate package, replacing this one; I have no other ideas at the moment)

**Comment by helhum** [comment]
Thanks for voicing your concerns. 

I like to give a little bit of perspective on the impact of switching to v4 for TYPO3 11 projects.

Most v11 installations will continue to work without change. The required changes in those which won't, will be minimal (make it work: <1h, fix thoroughly <1d).

The reason when something breaks is, hard coding a URL like `/typo3conf/ext/my_ext/Resources/Public/styles.css` in PHP, JS or CSS. Doing so is and in fact always has been usage of **internal** "knowledge" of where the extension is located in a path and how a URL is generated from that. Remember the fact, that all extensions up until now, theoretically can be installed in `typo3/ext/my_ext`)? Hard coding `/typo3conf/ext/my_ext/Resources/Public/styles.css` will break once the extension is installed globally in `typo3/ext/my_ext`. This is from TYPO3 version 3.8 throughout 11.x (and still with latest TYPO3 sources, installing globally, [still is an option](https://review.typo3.org/c/Packages/TYPO3.CMS/+/73626)).
I'm aware that not many people choose to use global extensions (or need a working state for both installation options).
Still, like with any usage of internal logic, doing so means (consciously or unconsciously) accepting it to break.

Now let's have a look at how many people will likely run into this issue during an update. 

First of all, when using Composer for installing any project, one needs to learn and understand what it means updating the dependencies of this project, what semantic versioning means, how Composer works with semantic versioning, how to explicitly update some dependencies, how to update transitive dependencies and what the impact of doing the above means. 
It is always balancing "staying up to date" with "stability". Therefore, I would *never* generally recommend to always use `--update-with-dependencies` flag. It depends on what one intends to do, while (ideally) understanding the impact of using the flag or not.

That said, please note that [helhum/typo3-secure-web](https://packagist.org/packages/helhum/typo3-secure-web) has a strict dependency to v3 of `typo3/cms-composer-installers`. This means, that everyone using this package as project dependency, will **never** get v4 installed. Judging from the installation count of this package, quite a lot of TYPO3 Composer installations are using it.

My conclusion is, that around 90-99% of v11 Composer installations won't be affected in a breaking manner with a v4 release of this package.

> a) make this possible breakage very well communicated upfront?

[Benni's article](https://b13.com/core-insights/typo3-and-composer-weve-come-a-long-way) is part of the communication strategy. We already received feedback from that. Feel free to suggest, or maybe just actually prepare and execute, further communication on this topic. The more we communicate about it, the better.

> b) make this totally opt-in in v11? (publish as separate package, replacing this one; I have no other ideas at the moment)

The question is: Do we want to have it opt-in or opt-out?

opt-in means: 
* We postpone the adoption (people starting with a new v11 project will get the old by default)
* We add yet another thing that needs to be done for those affected when upgrading to 12 (assuming we enforce v4 for 12)
* We don't get early feedback
* We optimise for 1%-10% of our (Composer) user base, instead of providing a future proof and more secure solution for 90%-99% by default

opt-out means:
* secure and future proof v11 installs by default
* early feedback
* complaints from some people from the 10%, when unconsciously upgrading one of their dependencies to a new major version

These are my thoughts and why I concluded to rather go for an opt-out in this case. I think it is the right choice, despite some complaints that will be voiced, even despite good communication on our end.

Last but not least: The time writing this, by far, exceeded the time required for anyone affected by the upgrade to v4 in some breaking way to fix their installation with `composer require typo3/cms-composer-installers: ^3` (or even better by finally at least using `helhum/typo3-secure-web`).

Still, I'm curious to hear arguments for a different strategy and for sure highly welcome any effort to communicate the current plans (opt-out).

**Comment by kevin-appelt** [comment]
We have already switched to v4 for testing purposes and have been doing quite well with it so far.
I like to share my experiences - maybe they help someone. ;)

As far as I know all our projects are currently equipped with helhum/typo3-secure-web, so nothing will automatically change for us with the v4 release. 
Most of the changes we had to make were probably “special cases” of our setup.

---

**Directory structure**
As mentioned we previously used `helhum/typo3-secure-web` and had `web/private` and `web/public` directories - with `typo3/cms-composer-installers v4`, `web/private` moved to `vendor` in the largest sense and we moved `web/public` to `public`. This of course meant that quite a few paths had to be adjusted in all sorts of areas (DDEV commands, deployment configuration, .gitignore, DDEV docroot, etc.).
If you have just used a `public` directory before you will have less and maybe even no work.

---

**Monitoring**
Our monitoring collects data (e.g. the version number) about the used extensions. In the largest sense we distinguished between “core extensions” and “other extensions”. The main distinction was (even if this is not 100% correct) the directory in which the extensions were located - `typo3/sysext` or `typo3conf/ext`. Similar to the TYPO3 Extension Scanner (which so far only scans `typo3conf/ext` - a [patch is already in the making](https://review.typo3.org/c/Packages/TYPO3.CMS/+/73679)) this is now no longer possible and had to be adapted accordingly. In our case the decision is now made by the vendor. We get the vendor information from the composer manifest of the package.

---

**Blocking unknown scripts**
To avoid loading JavaScript from extensions that may set cookies without the user’s consent, a middleware is used to block unknown script bindings that are not adapted for the Consent Management Platform. 
The previous solution was (in most cases) to “unblock” the scripts based on the src (e.g. `/typo3/sysext/adminpanel/Resources/Public/JavaScript/AdminPanel.js`). 
Unfortunately, the paths are now somewhat less identifiable like `_assets/<hash>/JavaScript/AdminPanel.js`.
I assume the hash doesn’t always remain identical, which effectively leaves only `JavaScript/AdminPanel.js` for identification. Pretty clear in that case. Let’s see if there are problems at some point when 2 extensions use the same filename in the same location inside their `Public` directory. But I’ll probably take care of that problem if the case really happens sometime.

---

In total, I spent about 3 hours on the changes - including reading the available information. And as I said I think most of the changes were necessary because of "special cases" of our setup.

**Comment by liayn** [comment]
Thank you Helmut for your extensive analysis.

We are well aware that this all is not much of a deal probably for people "near to the Core", like us three here on the ticket.
My thoughts are rather targeted towards those 80% of users, being neither close to Core nor present on Slack or anything like that.

The full story about composer and when to update what and when not and so on is for me a partial side story. Basically the thing here is that Core allows multiple major versions of a package, which can be beneficial or a problem.

Note on your secure-web package: Sorry to disappoint you, but we actually never used it and I am pretty damn sure that the majority of instances never heard of it. (which is a pity of course)

My suggestion in terms of communication would involve at least:

* News article on TYPO3.org
* Spamming of all social media channels, one week before release and on the release day
* Message on all still existing mailing lists

This communication must include the `composer require typo3/cms-composer-installers: ^3` command.

Please be aware that there are a lot of instances out there, where the owners are not willing to pay half a day of work to adjust for something "they don't need" and "they did not order". So it takes also time for agencies to get the message to the customers and convince them that this update is useful (and needs to be paid).

As outlined by Kevin there may be some dependencies on paths in CI pipelines or other advanced stuff, so this needs time to adjust as well.

These are may rather conservative thoughts on the impact. As always, I could see it too "dark" and in reality it will turn out just fine. The general direction is of course the right one, I never questioned that!

**Comment by liayn** [comment]
Thanks Helmut for sharing your thoughts. I of course did read the long version.
I won't steal you further time, hence I just reply to a few points.

Yes, I got to know about all that "by accident", because I try to follow some Slack channels and the blog link flew by. That was just pure luck.

**Communication channel**
I guess you know my opinion on sending out important information on social media (only).
So in terms of communication, the first place to look for (hopefully) true and more importantly trustworthy information should always be the company's/project's website. Hence, the news article on t3o. (Save your time feeding the other channels, that can be done by anyone)

**Why not me**
Important changes shall be announced by the people, who know the most about it and also deserve the honor. Me (and my particular low knowledge regarding composer installers) providing an article on t3o would be more like sarcasm.
In my opinion: If some bigger change is knowingly happening, it is also the duty of the authors to tell the story. Also in OSS.
(Otherwise we would have: "Oh yes, we changed this, but we didn't tell anybody upfront. We assumed people will discover and will tweet about it." I do not call this "the nice way".)

So if anyone involved with serious knowledge about the change (and the background story/ history) takes the time to writing down information, please please invest that time for a t3o news article and spare anything else.
Thanks.

(We can close this issue, if you like.)

**Comment by bmack** [comment]
@liayn @helhum so the only thing we need is a news article on t3o? Let's collaborate on this, I'd suggest :)

**Comment by kevin-appelt** [comment]
Quick update because I just thought of this issue again because of a conversation in Slack: We now have several projects running with v4 and no problems at all.
I think if the communication about the upcoming changes will be intensified and no one will be completely surprised by it, everything should go well. My suggestion would be to publish the article on t3o and share it as much as possible (at least on Slack - in #announcements and maybe even prepare a custom answer for it - and Twitter).
What is the status of the article for t3o?

**Comment by jonaseberle** [comment]
I see some work for integrators and developers that might not be trivial to solve.

One thing that will come up is "How to know an asset URL in JavaScript to include it?"

I for example am maintaining cookieman where in this example https://github.com/dmind-gmbh/extension-cookieman/blob/master/Configuration/TypoScript/TrackingObjects/Matomo.typoscript I need to know an URL (`.../example-inject.js`) in TypoScript which gets passed to JavaScript which outputs it to the page.
Currently I am not even aware how I could do that (probably some `FILE` magic or I might have to introduce an AJAX endpoint to fetch final URLs?).

Base line is that some extensions will be compatible with v11, but not with `typo3/cms-composer-installers:4.*` - good if someone checks it beforehand but most will just not notice. So of course there will be some frustration among users.

Many will run `composer update typo3/* -W` [as we have instructed them](https://docs.typo3.org/m/typo3/guide-installation/11.5/en-us/Minor/Index.html#execute-the-upgrade) and have regressions then.

It will be easier for the user base if `typo3/cms-composer-installers:4` would be TYPO3 v12 only, making it opt-in in v11.

**Comment by kevin-appelt** [comment]
@jonaseberle If an extension author depends on v3 it is possible to add it as a dependency.
`composer req typo3/cms-composer-installers:^3`
And so can everyone do, if one have serious problems with v4 - but I guess the problem is more to "work against the core instead of with the core".

**Comment by jonaseberle** [comment]
Sure. But this is idealistic.
It will just not be obvious which extensions will support it and which not because many extension authors will not expect having to do a compatibility check after an LTS version is released.

For me that tips the scale clearly towards not changing the CMSes directory structure within an LTS without an opt-in.

**Comment by jonaseberle** [comment]
Is there a way to still have predictable URLs like `public/_assets/my_ext/...` with `cms-composer-installers:^4`? Or is it imperative due to security concerns to have them scrambled?

**Comment by liayn** [comment]
@jonaseberle The way to generate URLs is usually something like this (and has always been):
`<div data-js-basepath="{f:uri.resource(path: ......)}">` 
JS is then using the well known data attribute to fetch what it needs to know.
(every framework actually works this way, you have some helper to generate the right paths)

**Comment by helhum** [comment]
> I see some work for integrators and developers 

I know, and I agree.

> that might not be trivial to solve.

There might be some things that are harder than other. I'd say, though that pretty much everything can be solved within hours rather than days.

> One thing that will come up is "How to know an asset URL in JavaScript to include it?"

Like [that](https://github.com/TYPO3/CmsComposerInstallers/issues/131#issuecomment-1084827893)

It requires a slightly different mindset, but once this is established, solutions for such things will come naturally.

> I for example am maintaining cookieman where in this example [https://github.com/dmind-gmbh/extension-cookieman/blob/master/Configuration/TypoScript/TrackingObjects/Matomo.typoscript](https://github.com/dmind-gmbh/extension-cookieman/blob/master/Configuration/TypoScript/TrackingObjects/Matomo.typoscript?rgh-link-date=2022-03-31T11%3A12%3A25Z) I need to know an URL (`.../example-inject.js`) in TypoScript which gets passed to JavaScript which outputs it to the page. 

The PHP code required to render a URL (or a script tag) [isn't too complex, I'd say](https://github.com/TYPO3/CmsComposerInstallers/issues/131#issuecomment-1084827893)

> EDIT: even more common than this example will be JavaScript module native `import` syntax (which will be used wildly from now on without RequireJS or any other lib needed) for importing modules cross-extension. EDIT2: This also affects dynamically loaded CSS as is the hype now around web components.

That is indeed the most complex thing to solve. TYPO3 has no modern concept for frontend code though yet. Depending on the individual solution used for bundling frontend assets, third party extensions can be used seamlessly or not, no matter which version of composer installers is used.

> Base line is that some extensions will be compatible with v11, but not with `typo3/cms-composer-installers:4.*` - good if someone checks it beforehand but most will just not notice. So of course there will be some frustration among users.
> Many will run `composer update typo3/* -W` [as we have instructed them](https://docs.typo3.org/m/typo3/guide-installation/11.5/en-us/Minor/Index.html#execute-the-upgrade) and have regressions then.

This is an argument, that was discussed already in this thread. I don't think it makes much sense to repeat them, but just to be sure:

`composer update typo3/* -W` is a risky operation in general. I can break your site in many ways.
Maybe this should be added to the documentation you referenced.

> It will be easier for the user base if `typo3/cms-composer-installers:4` would be TYPO3 v12 only, making it opt-in in v11.

I outlined the differences, advantages and disadvantages of this in this thread already. What exactly are your additional contribution to these? How are you weighing things differently, that you come to the conclusion opt-in is the better solution?

Again: This will affect "only" developers, performing a potentially risky operation. The fix for the "regressions" is a oneliner, that takes seconds to perform (given people are informed, which I assume here as we agreed already, that it is important enough to spread the word). Testing won't be affected (or can be opted out of), non Composer installation will of course also be not affected.

> These are some (most) verified TYPO3 extensions where some direct access to typo3conf/ext is used:

Even if most of the hits in the searches were relevant, how exactly would "opt-in for 11" solve that?
How would the extension owners (and TYPO3 core contributors) not need to wrap around their head around solving the issues that arise?
Even with "opt-in", if these extension are hard coding URLs or file system paths that contain `typo3conf/ext` for production (testing framework testes are not affected), they stay incompatible with composer installers 4

The only thing that happens with opt-in is, that we postpone everything for a couple of months/ years. As I pointed out: I think this would be a missed opportunity, traded for the "inconvenience" of performing a single composer operation that takes seconds.

Let's rather spend time on improving the documentation and finding solutions for a more secure, more composer compatible TYPO3 now.

**Comment by jonaseberle** [comment]
> Even if most of the hits in the searches were relevant, how exactly would "opt-in for 11" solve that?

It is common that extension authors will test compatibility with a new TYPO3 version. Most extensions (my test revealed 7 out of 10) should only allow `cms-composer-installers:^3` now. But I don't see that happening now before ^4 is released.

With `cms-composer-installers:^4` allowed in TYPO3 v11's `composer.json`, developers will install `cms-composer-installers:^4` in their TYPO3 v11 without really noticing when a stable v4 is released and might have functionality broken.

As @liayn wrote, achieving opt-in for v11 technically is not that easy now. To what Markus wrote I am adding the idea `composer req typo3/cms-composer-installers:"5.0.0 as 4.0.0"`, given that a stable v4 is not released and instead the next version would be 5.0.0.

----
The question about scrambled frontend URLs is more fundamental of course. Sorry I haven't noticed that change earlier. If it does not provide a huge security benefit, please consider `/_assets/my_ext/...` to allow frontend extensions to coexist more easily. Please think static CSS trying to load an image or font from another extension (`url('../../other_ext/Resources/Public')`).

**Comment by helhum** [comment]
@jonaseberle I understand you are concerned about such changes, that won't allow to hard code URLs, like it was possible for a few years. I'm listening and I try to explain to the best of my capacity, what exactly the changes are and why they make sense. 

In exchange I kindly ask you one last time, to show some respect, take given arguments into account and come up with appropriate counter arguments instead of repeating your opinion, which seem to mostly be based on assumptions. 

I'm fine with weighing arguments, but consider purely exchanging opinions a waste of time.

> > Even if most of the hits in the searches were relevant, how exactly would "opt-in for 11" solve that?
> 
> It is common that extension authors will test compatibility with a new TYPO3 version.

How does this answer my question? If you are already quoting it, could you be so kind also answering it? Thanks. 

> Most extensions (my test revealed 7 out of 10) should only allow `cms-composer-installers:^3` now. 

What exactly did you test? Which were the issues that surfaced? How did you choose the 10 and which exactly are the 10 and which one the 7? If you refer to the Github search, after navigating through the hits of 4 extensions, and did not find a single issue that would cause a bug, or would be fixable by just using API available for years (TYPO3 4.5 or earlier), I stopped looking any further. 

> But I don't see that happening now before ^4 is released.

Exactly! At some point, if we want to move further, we need to enforce usage of APIs

The only question that can be debated is when exactly and what the benefits and risks are enforcing them earlier by default in contrast to later. What exactly is your contribution to this question?
 
> With `cms-composer-installers:^4` allowed in TYPO3 v11's `composer.json`, developers will install `cms-composer-installers:^4` in their TYPO3 v11 without really noticing when a stable v4 is released and might have functionality broken.

So you are assuming that nobody will read announcements and the blindly make composer updates, push them to production without looking at the result?

One last time: Some people will be informed upfront, some have v4 in production without issues, some will notice that things are broken after update and will eventually find a solution that takes seconds to resolve,
some will be annoyed anyway. 
In fact, such things happen all the time, when doing even bugfix updates, the only difference here is, that we know it happens and can be prepared and even inform upfront. Incidentally, this very issue here is an information campaign, that got you informed. 

All that for me leads to the conclusion, that the risk is acceptable. 

If you have a different opinion, that's fine. If you want to contribute to the discussion though, you need to go further and come up with flaws in my arguments or bring up your own.

> As @liayn wrote, achieving opt-in for v11 technically is not that easy now.

It is absolutely possible. I haven't heard any arguments, how this would help though. 

> The question about scrambled frontend URLs

They are not scrambled. They are a hash of the relative path from composer root to the extension's public assets folder. In a default directory layout it's e.g. a hash of `vendor/typo3/cms-backend/Resources/Public`

Due to the current state of the TYPO3 code, there is no other way (I could come up with, challenge accepted) to derive a public URL form an absolute file system path. So in a given TYPO3 installation, these hashes never change. 

> please consider `/_assets/my_ext/...` to allow frontend extensions to coexist more easily. 

I considered this. Back then I thought it would not be possible. In retrospect, with some performance penalty, this would be possible, even in v11. 
But can you explain to me, how this helps making extensions compatible, that hard code typo3conf/ext?
It would help to publish to `typo3conf/ext/my_ext`. Is that your suggestion? Keep that URL for forever in order to never break anything in that area for years to come? Would be possible. We could abandon this approach discussed here, merge functionality of `helhum/typo3-secure-web` to the TYPO3 core and be done. 

I thought TYPO3 deserves a cleaner, more solid approach, where an API is introduced PHP side and frontend tooling side (JS/ CSS bundling), where extensions expose private and public resources alongside a custom URL for the latter. One step in that direction, would be to enforce using PHP APIs that exist and the step by step develop solutions for CSS/JS bundling with some recommended or default frontend toolset. 

> Please think static CSS trying to load an image or font from another extension (`url('../../other_ext/Resources/Public')`).

This is possible with either a frontend tooling (webpack, gulp, sass, grunt and friends), or by concatenation of JS/ CSS with TYPO3's concatenation features.

For dynamic ES module import, one would require a bundler, that can do code splitting and some further tooling around that.

But honest question: How many composer based TYPO3 projects have you done, where none of these options were available?
I might indeed be biased, as I've mostly seen bigger TYPO3 installations for bigger customers, but I haven't had a single one without at least some basic frontend tooling for years. 

And last but not least and most importantly:

What exactly is the impediment holding you back (and holding other maintainers of such sites back) to invest a minute to pin composer installers v3 to their installations?

In the meantime others can benefit from more secure TYPO3 installations by default, which will help as well everyone to slowly adopt a mindset of using APIs where ever possible and for us maintainers to catch more edge and use cases that challenges us to provide more solid APIs for future TYPO3 versions.

**Comment by helhum** [comment]
Let me clarify a few things after a night of sleep. The plan to release typo3 composer installers (t3ci) v4 with allowance for TYPO3 v11 (t3) was not lightheartedly made. For sure it is a tradeoff.

The idea is to trade a bit of instability and hassle for those, who don't follow communication channels, for better security and structure by default for new v11 instances.

There will be some site maintainers that don't want any hassle, which I totally understand. For those we offer to option to pin t3ci v3 and postpone any further hassle in that area until t3 v12 release, which will require t3ci v4 or higher.

There will be some extension maintainers, that don't want any hassle, which I understand as well. For those we offer guidance with encouraging to use existing APIs, the possibility it hard bin t3ci v3, if no other option is feasible and a lot of time until t3 v12 release to tweak some code and establish a new mindset. 

Let me stress here though, that compatibility with t3ci isn't a all or nothing thing. While some things might not work out of the box, for most issues it will be possible for site owners to still use such extension with making slight changes on their end. I'm not aware of any hard blockers, which would make any extension completely non functional with t3ci v4. I'm open to hear from concrete issues within third party extensions related to v4, which are (currently) impossible to solve or require more work than a few hours (if not minutes) to solve. 

Let me finish with outlining some technical detail, what I understand as "shift of mindset" that I consider necessary. 

Within TYPO3 code the boundary between an absolute file system path to a resource and a public URL is very loose. Most of the time they are used interchangeably, and one is derived from the other sometimes multiple times. Such code is brittle and can only work with the assumption, that all extension resources are publicly accessible by the web server (typo3conf/ext is in document root). 

When extension are now installed in the Composer vendor folder (which is the major feature of t3ci v4), this is not possible any more. We need to start thinking of URLs and file system paths as totally separate and distinct things and treat those accordingly in configuration and of course in code. 

Such mindset shift is quite radical and takes time. The earlier we start adopting it, the earlier it will be possible for TYPO3 to adopt a clean, modern, secure and fully Composer standard file structure.

Having t3ci v4 officially available for t3 v11 grants us more time, because we can start. 

Speaking of Composer standards, let me follow up on one last thing.

The fact that t3ci v4 is a major release, marks it conceptually as breaking and the therefore it factually *is* opt in, from a Composer point of view.
At the same time I totally understand, that for site owners, are used to yolo update their TYPO3 dependencies, because most of the time it works (and worked) out fine.
And because of that, I agree that it in this context it should rather be considered as opt-out to not install a not wanted version when updating all dependencies.

I hope I could help with providing some more details. I'm confident, that the new fully Composer standard file structure will benefit TYPO3 as a product and as a community. Let's make that happen with joining our efforts.

**Comment by jonaseberle** [comment]
I honestly feel a bit offended in being accused to not show respect or being too based on assumptions or having no experience with frontend tooling. Let me just say that you are very wrong in that respect. Thank you and others for your work and thoughts on this project! Your "last time" warnings have been received and understood.

**Comment by fsuter** [comment]
I stumbled on this issue when using TYPO3 Encore. A typical declaration could something like:
```
const config = Encore
    .setOutputPath('Resources/Public/Assets')
    .setPublicPath('/typo3conf/ext/catalog_package/Resources/Public/Assets')
    .setManifestKeyPrefix('Assets/')
    .addEntry('app', './Resources/Private/Assets/JavaScript/main.js')
    ...
```
which is JavaScript configuration for Webpack. In this case, a data attribute on some HTML tag cannot be used, since we are running in a pure JS context when compiling CSS and JS. How can I point to the right asset folder in this case?

For information, I have also directly reported this in: https://github.com/sabbelasichon/typo3_encore/issues/134

**Comment by avogt1701** [comment]
We also use TYPO3 Encore and had similar issues with composer installer v4. To avoid this problem we moved our public assets directory to "/public/build/".

**Comment by fsuter** [comment]
Ah, nice idea! Thanks.

**Comment by helhum** [comment]
> I stumbled on this issue when using TYPO3 Encore

Yupp. That is indeed something we need to solve before a stable release of CI4. At least give some very concrete advice how to handle that.

> To avoid this problem we moved our public assets directory to "/public/build/".

@avogt1701 I tried this as well (not using [`typo3_encore`](https://github.com/sabbelasichon/typo3_encore), but plain webpack) and ended up not being able to add the CSS and JS assets with TYPO3 `AssetCollector` API.

Are you like @fsuter using [`typo3_encore`](https://github.com/sabbelasichon/typo3_encore) and use its API to add the CSS and JS assets to the frontend output? From reading the code, this will then work out nicely. 

`AssetCollector` seems to be picky though and requires a filesystem path in the form of `EXT:ext_key/` to be able to add the assets correctly.

**Comment by helhum** [comment]
@jonaseberle 

> I honestly feel a bit offended in being accused to not show respect 

It wasn't and still isn't my intention to accuse you of anything. I only tried to explain *my own* boundaries regarding this very conversation. I'm sorry for choosing words, that let you feel offended.

> or having no experience with frontend tooling.

Oh, that is a complete misunderstanding. I of course have no idea what your experience is.
I was purely interested in, whether you happened to have stumbled upon TYPO3 projects, that use Composer, but use neither JS/CSS frontend bundling with the usual frontend tooling suspects, nor CSS/JS concatenation using the TYPO3 builtin CSS/JS concatenation mechanisms, thus just statically including CSS, that references assets from other extensions.

My assumption here is, that there aren't many and also that in such projects issues with CI4 could be fixed by choosing either of those options.

To not go too much off topic here, let's continue on Slack and post a summary here.

**Comment by avogt1701** [comment]
@helhum 
We include our assets via asset collector like this:

<f:asset.css identifier="app" href="build/Css/app.css"/>

To rewrite the final asset paths we use the "TYPO3\CMS\Core\Page\Event\BeforeStylesheetsRenderingEvent" event. This will rewrite the path "build/Css/app.css" to something like this "build/Css/app.6121f2d7.css" (mapping according to webpack manifest.json).

The internal TYPO3 logic will add the public path prefix to the final frontend output: ttps://github.com/TYPO3/typo3/blob/11.5/typo3/sysext/core/Classes/Utility/GeneralUtility.php#L2750

**Comment by ohader** [comment]
I'm using `"typo3/cms-composer-installers": "4.0.0-RC1"` in `composer.json` explicitly, without any problems in my TYPO3 v11 projects.

**Comment by helhum** [comment]
TYPO3 12 is released with dependency to v5 and v4 will stay available as rc for TYPO3 11 installations
